masschallenge/impact-api
# [AC-5087] - Improve previous-next mechanism in BaseListView in impact-api

**Changes Introduced in [AC-5087](https://masschallenge.atlassian.net/browse/AC-5087):**

- validate offset and limit values to be correct (offset is a non-negative integer, limit is a positive integer smaller than MAX_LIMIT=200).
- fix generation of previous/next urls around edge cases (close to offset==0 or offset==count)

**Test:**

- glossary:
  - limit - the number of results returned in the response ("results per page").
  - offset - the index of the element that begins the response window ([offset:offset+limit])
  - count - the total number of results in a request (NOT a parameter)


- Setup:
  - `make dev; 
make load-remote-db DB_FILE_NAME=initial_schema.sql.gz; 
make grant-permissions PERMISSION_USER=test@example.org PERMISSION_CLASSES=v1_clients`
  - [Create an application](http://localhost:8000/oauth/applications/register/) for the user with the permissions.
  - Load [Postman collection](https://www.getpostman.com/collections/0a95773f88df764b15ee), initialize the token
  - in V1 folder, open v1 user list GET
  - make the default call, see that it works, and is the same as development
- Test:
  - set the query parameters ?limit=20 . Same as development.
  - set the query parameters ?limit=10&offset=5 . On development, "previous" would have a negative offset. On branch it does not (will be have an implicit offset=0).
  - set the query parameters ?limit=10&offset=15. Should be the same as development.
  - set the query parameters ?limit=10&offset=47915 (count=47920). "next" will be null on branch.
  - set the query parameters ?limit=10&offset=50000 (count=47920). "previous" will be 47910(= count-limit) on branch.
  - Go to "v1 program family list GET". set the query parameters ?limit=20 (count=11). "next" and "previous" should both be null on branch.
  - set the query parameters ?limit=300. See error on branch.
  - set the query parameters ?limit=5.5. See error on branch.
  - set the query parameters ?limit=0. See error on branch.
  - set the query parameters ?limit=-5. See error on branch.
  - set the query parameters ?limit=foo. See error on branch.
  - set the query parameters ?offset=5.5. See error on branch.
  - set the query parameters ?offset=-5. See error on branch.
  - set the query parameters ?offset=foo. See error on branch.
  - set the query parameters ?offset=0. The same as not adding the parameter, successful.
